### ISSUE TITLE
Geneate silence at the beginning of a clip creates an inconsistent, unnecessary, split

### ISSUE THREAD

**Issue by petersampsonaudacity** [audacity/audacity#1653]
**Describe the bug**
When generating silence at the beginning of an audio clip two clips are created from the audio with a clip boundary at the end of the generated silence,  The steps show this starting at T=0,   but the same is tue if the clip starts at a later time and if the clip is not the first clip in the track 

This is inconsistent with Generate > Chirp, DTMF tones, Noise and Tone  -  which leave the audio as a single clip.

This does not happen if you generate silence in and aup to the end of a clip, or if you create silence in the middle of the clip - also an inconsistency.

**To Reproduce**
1. generate 30 second chirp
2.  select from 0-10 seconds
3.  Generate silence
4.  Observe: the silence is generated but a clip boundary is created at 10 seconds.

**Expected behavior**
No clip boundary should be created - the audio should remain as a single clip

**Screenshots**
None

**Additional information (please complete the following information):**
 - OS: Windows 10 - but assume all OS
 - Version:  Audacity 3.1.0

**Additional context**
This creates an additional problem with respect to 
 Generate silence creates a clip with no clip name (in the Clip-handle) #1651 

Because, if after step 4 the user decides they don't want the split and then joins to the split clips, the resulting consolidated clip has no name and the original clip name is lost.

**Comment by petersampsonaudacity** [comment]
The same is true for:
a) Generate > Pluck,
b) Generate > Rhythm Track 
c) Generate > Rissset Drum

which also create two unnecessary clips from the audio when generatyin at the start of a clip bit not at the end or in the middle of a clip

**Comment by micsthepick** [comment]
I would like to take a look at this issue, it seems to be something that I might be able to understand.

**Comment by LWinterberg** [comment]
alright, I'll assign it to you for a while. Keep us updated on how progress is going :)

**Comment by micsthepick** [comment]
ok, looking in to it this evening, and I'll see if I can make some progress on it tomorrow.

**Comment by micsthepick** [comment]
do we want to mirror the current behavior of chirp completely? For example, if the selected region is entirely inside a track (not just aligned to the start), it will overwrite the contents, and if it extends out from either end, a new clip is created to contain the newly generated audio.

**Comment by micsthepick** [comment]
It should be easy enough to look at the code for "silence audio section", and use that as an example of how to process the effect without creating a new clip.

**Comment by micsthepick** [comment]
except that it will need to also add clips where there are no clips present

**Comment by micsthepick** [comment]
ok, so I've identified the issue, Generator.cpp calls ClearAndPaste passing in the argument merge=False, where as Effect.cpp calls the same method with merge=True. The question is, would there be any unintended side effects for making Generator.cpp pass merge=True?

**Comment by SteveDaulton** [comment]
> 1. generate 30 second chirp
> 2. select from 0-10 seconds
> 3. Generate silence
> 4. Observe: the silence is generated but a clip boundary is created at 10 seconds.

I don't see the clip  boundary in Audacity 3.2.0-alpha-20211230 (Debug) on Linux:

![First Track000](https://user-images.githubusercontent.com/11681097/147796842-6e88ba38-99ad-4ec0-b7a0-4875ca972c31.png)

**Comment by micsthepick** [comment]
@SteveDaulton are you sure that you are using Generate -> silence instead of silence selection?

**Comment by micsthepick** [comment]
double checked, and the problem is still there as of 148213d1851b671dba062a13b31b1b42915e8ff4 at least for windows

**Comment by petersampsonaudacity** [comment]
Testing on W10 with 3.1.3  I can no longer reproduce this issue:

![image](https://user-images.githubusercontent.com/50205984/147818346-1c3d907d-f5a4-4cc4-9825-4225a3601696.png)

In fact I can't seem to reproduce it now on 3.1.0 - the version I originally logged this against.

**Comment by SteveDaulton** [comment]
> @SteveDaulton are you sure that you are using Generate -> silence instead of silence selection?

Yes I'm using "Generate -> Silence".

Note that in recent versions of Audacity, there has been a change in "Preferences > Tracks > Track Behaviors".
"Editing a clip can move other clips" is now off by default (a bad decision imo, but that's another story). If I enable that option, then I see the behavior that you describe.

**Comment by petersampsonaudacity** [comment]
Ah tes, if I turn that option back on then I do get the clip when I follow my earler steps to reproduce (BTW this explains why 3.1.0 worked for me now, as that option was "off")

But it's only if I creat a silence that starts from T=0
![image](https://user-images.githubusercontent.com/50205984/147824481-d3b19b78-8d12-4055-88c5-2e2df3255b48.png)

**Comment by petersampsonaudacity** [comment]
Or rather if I generate silence from the beginning of the clip
![image](https://user-images.githubusercontent.com/50205984/147824550-3762c6e4-1c0f-421d-a9d8-e126a3087e6b.png)

**Comment by petersampsonaudacity** [comment]
But basically not only is it wrong in the first place, it's grossly inconsistent.

**Comment by SteveDaulton** [comment]
I'm not actually convinced that this is a bug. "Silence" is a special case of "audio", so it doesn't seem unreasonable for it to be handled as a special case. While I love consistency, I don't agree to being a blind slave of consistency. I think we should be asking which behavior is more useful, rather than whether generating silence must be consistent with generating non-silence.

**Comment by petersampsonaudacity** [comment]
And now look what happens if I start th selection for the Generate silence before the start of the clip the gray space
![image](https://user-images.githubusercontent.com/50205984/147824709-cc396dea-40d1-492d-9334-129eb5949b67.png)
![image](https://user-images.githubusercontent.com/50205984/147824714-bed17e23-ea12-476a-980b-9a43d1523314.png)
The new clip from 5-15 secs is probably justified - but the one at 25 is really not (as per this bug)

**Comment by petersampsonaudacity** [comment]
BTW using Ctrl+L with that move clips option turned on does NOT create a nee clip
![image](https://user-images.githubusercontent.com/50205984/147824842-99d1995f-ba56-4cf9-946a-2fb1ad4427a1.png)

**Comment by petersampsonaudacity** [comment]
> I'm not actually convinced that this is a bug. "Silence" is a special case of "audio", so it doesn't seem unreasonable for it to be handled as a special case. While I love consistency, I don't agree to being a blind slave of consistency. I think we should be asking which behavior is more useful, rather than whether generating silence must be consistent with generating non-silence.

The inconsistency is that  you only get a split **_if_** the selection of part of the clip starts at the beginning (or before) the start of the clip.  If you are tom get clips made like this then surely selecting a chunk in the middle to silence should cause 3 clips to be made.

A second inconsistency is that it depends on the setting foe "Editing a clip can move other clips.  That subtle difference is undeniably not very "discoverable" - _I certainly didn't discover it when I was testing, I needed you, Steve, to point it out.   ;-)_

A third  inconsistency is that Ctrl+L behaves differently _(this may be an acceptable difference)_.

**Comment by SteveDaulton** [comment]
> A third inconsistency is that Ctrl+L behaves differently _(this may be an acceptable difference)_.

I think that "Ctrl + L" should work like an "Effect" rather than like a "Generate" (and I think it does).

**Comment by SteveDaulton** [comment]
> The inconsistency is that you only get a split **_if_** the selection of part of the clip starts at the beginning (or before) the start of the clip.

In Audacity 2.0.0, **_all_** of the Generate effects did that.
It seems that someone changed the behavior for _most_ of the Generate effects, but missed "Generate Silence". They also messed up with Nyquist Generate effects - if the selection overlaps the end of a clip, the split is in the wrong place.

**Comment by petersampsonaudacity** [comment]
> It seems that someone changed the behavior for _most_ of the Generate effects, but missed "Generate Silence". They also messed up with Nyquist Generate effects - if the selection overlaps the end of a clip, the split is in the wrong place.

Bo they missed it for all the below the line shipped generators, the Nyquist generators.

**Comment by petersampsonaudacity** [comment]
I have just tested  @micsthepick 's branch build for this issue on W10
https://github.com/audacity/audacity/actions/runs/1639516074

This fix does fix the use case for "Silence" 

Bit it does NOT fix it for any of the three Nyquist generators
a) Pluck
b) Rhythm Track
c) Risset Deum

![image](https://user-images.githubusercontent.com/50205984/147851275-186862d3-4f49-4ac2-8511-c93656c4486c.png)
